Two small changes to make by default, OOTB, work better with msys2.

xeiavicaflashrepository opened the pull request:
msys2 is a rather popular semi-fork of cygwin to bring linux software to microsoft windows.

I made two tiny additions to make it work OTTB better with windows without having to patch it ourselves.

First one is if $HOSTTYPE is detected as msys2, do something differently from cygwin because of how symlinks work in msys2 compared to cygwin. Second one is to add a shebang to the start of src/cmd/INIT/mamprobe.sh because windows cannot load the file without sed-ing the shebang into it. This shouldn't make a difference because it's already a shell script to begin with.

I would like to see better type and HOSTTYPE detection with msys2. Let me know your thoughts and how I can change the patch to be better fitting for ksh93 and msys2.

Commit:
These two changes are necessary to work with msys2 on windows because of how the operating system works.

xeiavicaflashrepository commented:
Also, I should add, there already exists someone with problems with msys2. This should help fix things.
https://github.com/ksh93/ksh/issues/801

jeremyd2019 commented on bin/package:
```diff
@@ -1790,6 +1790,11 @@ then	case $(uname) in
 	esac
 fi
 
+# Fix for msys2 because /bin is the same thing as /usr/bin.
+if test /bin -ef /usr/bin; then
+	DEFPATH="/usr/bin:$DEFPATH"
+fi
```
I'm concerned this may be the wrong thing, ie for Cygwin, which does the opposite as MSYS2 (it bind mounts /usr/bin to point to /bin).  Could we check HOSTTYPE at this point in the script?

```suggestion
# Fix for msys2 because /bin is the same thing as /usr/bin.
case $HOSTTYPE in
msys.*)
	if test /bin -ef /usr/bin; then
		DEFPATH="/usr/bin:$DEFPATH"
	fi
esac
```

jeremyd2019 reviewed (commented):
Are there other places that know HOSTTYPE of `cygwin.*` that need to also check for `msys.*`?

ghost commented:
I want to add, that the files `src/lib/libast/path/pathposix.c` and `src/lib/libast/path/pathnative.c` use a function that has been deprecated and removed a long time ago. GNU ldd reports back a linking error. I fixed by commenting it out which I'll remove in my PR.

Commit:
Removed two obsolete and long removed functions breaking linkage.

jeremyd2019 commented:
If you want to convert paths, you would use [`cygwin_conv_path`](https://cygwin.com/cygwin-api/func-cygwin-conv-path.html)

Commit:
Reverted previous commit. Nothing was wrong if the files, it was the fact it only pulled in sys/cygwin.h IF it detected as cygwin. Added an || statement to let it check if it's __MSYS__ as well. More will probable have to be changed.

xeiavicaflashrepository commented:
@jeremyd2019 
Get this, there was nothing wrong with the compiler script nor ksh93. It was windows itself that was the problem (What else is new). There is a for loop that starts with `for i in ksh` and what this would do is set the default shell to ksh. Windows didn't like that and it is why it resulted in a fork bomb of sorts. I'll close my other issue since I know it's not the script to blame. 

I simply moved sh to the front, and it works now. As for the two files I modified, it would definitely break something if I removed them. They were merely deprecated but not removed. That's why it worked IF we set it to cygwin, because the if macro that pulls in sys/cygwin.h is ONLY if __CYGWIN__ is detected. I added in an || statement to include __MSYS__.

jeremyd2019 commented:
On Fri, 14 Mar 2025, xeiavicaflashrepository wrote:

> I simply moved sh to the front, and it works now. As for the two files I
> modified, it would definitely break something if I removed them. They
> were merely deprecated but not removed. That's why it worked IF we set
> it to cygwin, because the if macro that pulls in sys/cygwin.h is ONLY if
> __CYGWIN__ is detected. I added in an || statement to include __MSYS__.

MSYS2 gcc still sets __CYGWIN__ in addition to __MSYS__ so there shouldn't
have been any problem.

jeremyd2019 commented:
There is no `cygwin_conv_to` in Cygwin codebase anymore.  I still think you want to use `cygwin_conv_path`, but apparently this code isn't used because it should fail to link.  BTW, don't know what you meant by `GNU ldd`.  Did you mean `ld`? Or did you mean Cygwin's `ldd`?

Commit:
Fixed an issue with compilation.

jeremyd2019 commented:
No.  That code would fail on modern Cygwin too.  Should be changed to use `cygwin_conv_path` instead.

Commit:
Hopefully this time things turn out well.

jeremyd2019 commented on src/lib/libast/comp/omitted.c:
```diff
@@ -1007,7 +1008,7 @@ ctime_now(const char* path)
 	if (sysstat(path, &fs) || (fs.st_mode & S_IWUSR) || syschmod(path, (fs.st_mode | S_IWUSR) & S_IPERM))
 		fs.st_mode = 0;
-	cygwin_conv_to_win32_path(path, tmp);
+	cygwin_conv_path(CCP_POSIX_TO_WIN_W | CCP_ABSOLUTE, path, tmp, 0);
```
```suggestion
	cygwin_conv_path(CCP_POSIX_TO_WIN_A | CCP_ABSOLUTE, path, tmp, sizeof(tmp));
```
Similar for the other calls to cygwin_conv_path.

No idea where conv_path_list_buf_size (or  conv_last_list_buf_size) come from.  But if you need to convert a path list, you can use [cygwin_conv_path_list](https://cygwin.com/cygwin-api/func-cygwin-conv-path-list.html)

jeremyd2019 replied:
so it looks like these old deprecated functions were 32-bit only, and removed in cygwin/cygwin@3e917daec121ae5fd5cdcbb25d0d4e9d1cfd0a2a

jeremyd2019 commented:
This code does not seem to be enabled on (current) Cygwin because the test in src/lib/libast/features/omitted tries to use functions prefixed with `_` and Cygwin does not define them.  If I define them all to the non-underscore variant, the test produces:
```c
#define _win32_botch_copy       1
#define _win32_botch_rename     1
#define _win32_botch_execve     1
#define _win32_botch    1
```

I added include of `<stdio.h>` and:
```c
#ifdef __CYGWIN__
#define _open open
#define _fstat fstat
#define _fstat fstat
#define _read read
#define _write write
#define _close close
#define _mkdir mkdir
#define _chdir chdir
#define _access access
#define _chmod chmod
#define _getpagesize getpagesize
#define _link link
#define _pathconf pathconf
#define _rename rename
#define _stat stat
#define _truncate truncate
#define _unlink unlink
#define _utime utime
#endif
```

But if this code is enabled, it is almost certainly going to be really broken.

jeremyd2019 commented:
I think maybe it's better that those "botch" patches are not enabled, Cygwin does those behaviors for a reason
